### ISSUE TITLE
Rate Limiting may block adding a GA account

### ISSUE THREAD

**Issue by emilong** [metabase/metabase#3912]
- Metabase version: 0.21.0
- Metabase hosting environment: Elastic Beanstalk
- Metabase internal database: Google Analytics

Upon adding my Google Analytics account as a database, MB hit the Google API rate limit almost immediately and was not able to sync it as a DB. I can't find the logs for this (I had to spin up a new container), but it was pretty straightforwardly printed as an exception in MB's logs (obtained via `docker logs` in my Beanstalk env). The UI did not show any issues.

One inciting factor may be that my Google Analytics account has 10 properties (i.e. tracking codes), which may have caused MB to need more API calls than an account with fewer. (?) I'm not using the Analytics API in any way other than MB (i.e. only MB was able to cause hitting the rate limit).

I was able to workaround this (see below), but:

1. Some UI would be nice for this, though I'm guessing its pretty asynchronous. In comparison, the error UI around setting up the OAuth was austere, but definitely got the job done. I was able to discover that I was copying the wrong string for the account ID and that I needed to enable the Analytics API in the Google Dev console as a relative noob wrt to Google APIs.
2. I'm not sure if I hadn't bumped the quota that MB would have ever completed. I.e. if the initial sync is atomic and rolls back as soon as it hits the rate limit, would that have prevented any attempt to sync?

### Workaround

1. Visit https://console.developers.google.com/
2. Find your project
3. Search for "All Quotas" in the search box
4. Edit all the quotas for Analytics API, bumping them to their max allowed (I only had to/could bump Queries per 100 seconds to 2000)

**Comment by tlrobinson** [comment]
@emilong To confirm, it's hitting the requests/second limit, not the requests/day limit?

**Comment by tlrobinson** [comment]
(Also FYI admins can see the backend logs in the UI here: 
![screenshot 2016-12-09 11 46 22](https://cloud.githubusercontent.com/assets/18193/21062314/24e4a7ee-be05-11e6-89af-f21184f9b51d.png)

**Comment by emilong** [comment]
@tlrobinson yup, I believe it was hitting the requests/second limit as increasing that (on the same day) fixed the issue.

Thanks for the tip on the logs! They don't appear to go back far enough for the exception. LMK if there's a way to paginate backwards, but even curl'ing the endpoint and looking through that way didn't get to it.

**Comment by Poggen** [comment]
Issue still exists on 0.22 – just updated and then tried to add my GA database:
```
01-10 22:45:22 INFO driver.google :: Fetching Google access/refresh tokens with auth-code '4/xZ5XgbS5DBGqZkBgoQPvyViOL01WqG9eBPQXqhd-iAw'...
01-10 22:45:27 ERROR metabase.driver :: Failed to connect to database: Timed out after 5000 milliseconds.
01-10 22:45:27 DEBUG metabase.middleware :: POST /api/database 400 (5 s) (0 DB calls)
{:valid false, :dbname "Timed out after 5000 milliseconds.", :message "Timed out after 5000 milliseconds."}
```
Any ideas on how to get around it? Maybe worth mentioning that we have multiple GA accounts with a total of 50+ properties in total. The account I tried adding now has 50 out of a total of 50 allowed properties per account.

**Comment by Poggen** [comment]
Issue persists on 0.22. Just tried to add one of our GA accounts (50/50 properties) to a freshly updated metabase, but got the following error:
```
01-10 23:42:33 ERROR metabase.driver :: Failed to connect to database: Timed out after 5000 milliseconds.
01-10 23:42:33 DEBUG metabase.middleware :: POST /api/database 400 (5 s) (0 DB calls)
{:valid false, :dbname "Timed out after 5000 milliseconds.", :message "Timed out after 5000 milliseconds."}
```
Please see enclosed screenshot of our google api quota history/settings:
![image](https://cloud.githubusercontent.com/assets/5216834/21826472/e54aa2c2-d787-11e6-81a7-9b4a4295fbfa.png)

**Comment by Hennisa** [comment]
Getting the same error "Timed out after 5000 milliseconds" when trying to add GA as a new data source. 
Also, please see enclosed screenshot of our google api quota:

<img width="896" alt="screen shot 2017-02-05 at 21 55 16" src="https://cloud.githubusercontent.com/assets/14327271/22629817/d6f3da16-ebed-11e6-8db3-db48cb4a2f35.png">

**Comment by salsakran** [comment]
@Hennisa how many web properties do you have on that account?

**Comment by alextobies** [comment]
We have the same problem as described by @Hennisa.
@salsakran Our account holds 36 properties with 2 views each.

**Comment by VikramTiwari** [comment]
@salsakran I recently encountered this with an account with more than 100 properties.

Another issue:
During hourly sync, the per-second limits are easily crossed over if the number of properties connected using a single account is large in number. Though this could easily be sorted by using different auth credentials for each property, yet it is an issue.

![screen shot 2017-04-18 at 16 13 10](https://cloud.githubusercontent.com/assets/1330677/25157435/522cdffc-2456-11e7-8440-cfb87937abe6.png)

I am also interested in knowing if folks see this happen to their traditional databases? Like if they have a large number of tables in their DB do they see spikes during hourly syncs and how do they handle it there?

One of the possible ways Metabase can handle this is by using lazy sync where the sync is deliberately made slow. But this could cause issues when the time taken to sync all databases is same or greater than the interval of sync. Thoughts? @camsaul

**Comment by alextobies** [comment]
Any updates on this?

What about changing the API access from the account to the property level?

**Comment by Poggen** [comment]
Agree with @alextobies that an update regarding this would be nice. I have tried adding an account with fewer properties and had no issues. But when I try to add our primary account with 50 properties it still fails with the "Timed out after 5000 milliseconds" error. Here's an overview of the google api traffic by method when I try to add the account with 50 properties in metabase. It looks to me as if it is the `analytics.management.profiles.list` that spikes the quota, as it only fires 50 calls (the other 9 calls is from another account).
<img width="937" alt="skarmavbild 2017-07-07 kl 16 30 56" src="https://user-images.githubusercontent.com/5216834/27962296-c1a8a644-6331-11e7-8f0f-8eb4dc400468.png">

**Comment by lawrencedudley** [comment]
We're seeing this error as well on accounts with many profiles

**Comment by mynock** [comment]
Can we increase the `can-connect-timeout-ms` default or make it configurable? I think that should help.

**Comment by camsaul** [comment]
We could make that configurable, but I feel like that is putting a band-aid on the problem rather than fixing it. It would be better if we could make Metabase a little smarter so it issues less requests and doesn't hit the rate limit in the first place

**Comment by mynock** [comment]
I totally understand, but perfect may be the enemy of good here as well. I'm not sure it was actually hitting the rate limit in the first place since our limits are as high as we can make them without requesting more from Google. I forked the code and upped the timeout a bit and was able to add our GA data after deploying that fork. So even if this is a band-aid it might be worthwhile to put it on until there's time to do the actual surgery. I think it probably took around 6-7 seconds to add our GA account (with 33 properties) on 0.26

**Comment by camsaul** [comment]
Fair enough. I'll look at adding a configuration option here

**Comment by camsaul** [comment]
Added a new `MB_DB_CONNECTION_TIMEOUT_MS` env var in #5899. That will let you manually configure the value as needed @mynock 

That will be available as part of our upcoming 0.26.0 release. We're still looking at a way to fix this problem nicely

**Comment by camsaul** [comment]
Closing as not planned due to dropping GA support
